src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -47,6 +47,36 @@ public SocketInformation DuplicateAndClose(int targetProcessId)
             Debug.Assert(!_handle.LastConnectFailed);
         }
 
+        private static unsafe void LoadSocketTypeFromHandle(
+            SafeSocketHandle handle, out AddressFamily addressFamily, out SocketType socketType, out ProtocolType protocolType, out bool blocking)
+        {
+            // Validate that the supplied handle is indeed a socket.
+            if (Interop.Sys.FStat(handle, out Interop.Sys.FileStatus stat) == -1 ||
```

Comment:
could we skip this? I think not would be useful in several cases like reading data from character device or /proc on Linux. Current System.IO does not really have Async mode or concept of data availability. Using select/poll or Async IS would be great match IMHO. I can also do more testing and follow up on this if you don't want to deal with it now.

Reply:
> Using select/poll or Async IS would be great match IMHO

Yes, though I don't know that encouraging use of `Socket` is the right abstraction for that.

> I can also do more testing and follow up on this if you don't want to deal with it now.

If we can make everything work, I'd be ok with removing the constraint, but I was hitting hangs otherwise, in particular as part of socket shutdown, if I allowed it to be used with file descriptors other than sockets.  You might have better luck, or maybe there are just some small tweaks we need to do in the implementation.  I'm going to leave this in for now, though.